src/modules/roc_core/target_posix/roc_core/time.cpp

```diff
@@ -49,6 +52,117 @@ void sleep_for_ms(uint64_t ms) {
         }
     }
 }
+#elif defined(__APPLE__) && defined(__MACH__)
+
+#include <mach/clock.h>
+#include <mach/clock_types.h>
+#include <mach/mach.h>
+#include <mach/mach_time.h>
+
+#define CLOCK_MONOTONIC SYSTEM_CLOCK
+
+uint64_t timestamp_ms() {
+    clock_serv_t cclock;
+    mach_timespec_t mts;
+
+    // Receive a permission to send a messages to a specified clock service.
+    kern_return_t ret =
+        host_get_clock_service(mach_host_self(), CLOCK_MONOTONIC, &cclock);
+    if (ret != KERN_SUCCESS) {
+        roc_panic("host_get_clock_service: %d", ret);
+    }
+
+    // man for a clock_get_time promises that it increments monotonically.
+    // https://opensource.apple.com/source/xnu/xnu-2422.1.72/osfmk/man/clock_get_time.html
+    //
+    // But if you really want to call clock_set_time, kernel will return an
+    // error for it.
+    // https://opensource.apple.com/source/xnu/xnu-2422.1.72/osfmk/man/clock_get_time.html
+    ret = clock_get_time(cclock, &mts);
+    if (ret != KERN_SUCCESS) {
+        roc_panic("clock_get_time(CLOCK_MONOTONIC): %d", ret);
+    }
+
+    ret = mach_port_deallocate(mach_task_self(), cclock);
+    if (ret != KERN_SUCCESS) {
+        roc_panic("mach_port_deallocate: %d", ret);
+    }
+
+    return uint64_t(mts.tv_sec * 1000 + ((long)mts.tv_nsec) / 1000000);
+}
+
+void sleep_until_ms(uint64_t ms) {
+    mach_timespec_t ts;
+    ts.tv_sec = (unsigned int)ms / 1000;
+    ts.tv_nsec = ms % 1000 * 1000000;
+
+    kern_return_t ret = KERN_SUCCESS;
+    for (;;) {
+        // We are interested in using SYSTEM_CLOCK (aka CLOCK_MONOTONIC in
+        // other world), digging into XNU source code we can find that if we
+        // set a name for a clock port to MACH_PORT_NULL, a kernel will use
+        // SYSTEM_CLOCK.
+        //
+        // https://opensource.apple.com/source/xnu/xnu-2422.1.72/osfmk/kern/clock_oldops.c
+        // We are interested in a @clock_sleep_trap, because it's used by
+        // @clock_sleep under the hood.
+        ret = clock_sleep(MACH_PORT_NULL, TIME_ABSOLUTE, ts, NULL);
+        if (ret == KERN_SUCCESS) {
+            break;
+        }
+
+        if (ret != EINTR) {
```

Comment:
Not, clock_sleep doesn't return EINTR (4), it returns KERN_ABORTED (14), this mistake was introduced because previously I received EINTR from os x ipc subsystem during receiving rights to mach port (no need to go deeper in this concept).

I've fixed it.